plutus-emulator/src/Wallet/Emulator/NodeClient.hs

```diff
@@ -0,0 +1,136 @@
+    , vlbRest   :: [Tx]
+    -- ^ The transactions that haven't been validated because the current slot is
+    --   not in their validation interval.
+    }
+
+-- | A pool of transactions which have yet to be validated.
+type TxPool = [Tx]
+
+data ChainState = ChainState {
+    _chainNewestFirst :: Blockchain, -- ^ The current chain, with the newest transactions first in the list.
+    _txPool           :: TxPool, -- ^ The pool of pending transactions.
+    _index            :: Index.UtxoIndex -- ^ The UTxO index, used for validation.
+} deriving (Show)
+
+makeLenses ''ChainState
+
+-- | Validate a block given the current slot and UTxO index, returning the valid
+--   transactions, success/failure events, remaining transactions and the
+--   updated UTxO set.
+validateBlock :: Slot -> Index.UtxoIndex -> [Tx] -> (ValidatedBlock, Index.UtxoIndex)
+validateBlock slot@(Slot s) idx txns =
+    let
+        -- Select those transactions that can be validated in the
+        -- current slot
+        (eligibleTxns, rest) = partition (canValidateNow slot) txns
+
+        -- Validate eligible transactions, updating the UTXO index each time
+        (processed, idx') =
+            flip runState idx $ for eligibleTxns $ \t -> do
+                r <- validateEm slot t
+                pure (t, r)
+
+        -- The new block contains all transaction that were validated
+        -- successfully
+        block = fst <$> filter (isNothing . snd) processed
+
+        -- Also return an `EmulatorEvent` for each transaction that was
+        -- processed
+        nextSlot = Slot (s + 1)
+        events   = SlotAdd nextSlot : (uncurry mkValidationEvent <$> processed)
+
+    in (ValidatedBlock block events rest, idx')
+
+-- | Check whether the given transaction can be validated in the given slot.
+canValidateNow :: Slot -> Tx -> Bool
+canValidateNow slot tx = Interval.member slot (txValidRange tx)
+
+mkValidationEvent :: Tx -> Maybe Index.ValidationError -> ChainEvent
+mkValidationEvent t result =
+    case result of
+        Nothing  -> TxnValidate (txId t)
+        Just err -> TxnValidationFail (txId t) err
+
+-- | Validate a transaction in the current emulator state.
+validateEm :: MonadState Index.UtxoIndex m => Slot -> Tx -> m (Maybe Index.ValidationError)
+validateEm h txn = do
+    idx <- get
+    let result = Index.runValidation (Index.validateTransaction h txn) idx
+    case result of
+        Left e -> pure (Just e)
+        Right idx' -> do
+            _ <- put idx'
+            pure Nothing
+
+class Monad m => NodeClientAPI m where
+    publishTx :: Tx -> m ()
+    currentSlot :: m Slot
+
+-- | Pure implementation of the NodeClient API
+
+instance NodeClientAPI (State ChainState) where
+    publishTx tx = state $ ((), ) . over txPool (tx :)
+
+    currentSlot =
+        gets (Slot . fromIntegral . length . _chainNewestFirst)
+
+processBlock :: MonadState ChainState m => m ValidatedBlock
+processBlock =
+    state $ \st ->
```

Comment:
I'd mildly prefer using `get/put` over `state`, if nothing else just to avoid the extra indentation...

Reply:
There is a problem here if we think this will eventually be a TVar where we need to have an atomic read/update. If we use get/put we would not be able to properly implement it atomically.

Reply:
I moved the `state $ st ->` after the `=` and now the indentation is avoided.

Reply:
Eh, we can change it to be `state` or `modify` when we have the `TVar`. Generally for stuff like this I think it's easier to just change it when we get there. But not a big deal.